e2e-testing/readme.md

```diff
@@ -0,0 +1,164 @@
+
+### **3) Ranorex Studio**
+
+![Ranorex Studio](/images/ranorex-studio2.png)
+
+**Ranorex Studio** is a complete end-to-end test automation tool for desktop, web, and mobile applications. Create reliable tests fast without any coding at all, or using the full IDE. Use external CSV or Excel files or a SQL database as inputs to your tests.
+
+Run tests in parallel or on a Selenium Grid with built-in Selenium WebDriver. Ranorex Studio integrates with your CI/CD process to shorten your release cycles without sacrificing quality.
+
+**Ranorex Studio** tests also integrate with Azure DevOps (AzDO), which can be run as part of a build pipeline in AzDO.
+
+=> [Visit Ranorex Studio Website](https://www.ranorex.com/ranorex-studio-test-automation/?utm_source=softwaretestinghelp&utm_medium=cpc&utm_campaign=softwaretestinghelp_what-is-end-to-end-testing) or read about its [integration with AzDO](https://www.ranorex.com/help/latest/interfaces-connectivity/azure-devops-integration/introduction/)
+
+### **4) Katalon Studio**
+
+![Katalon](/images/New-Logo-Katalon-Studio.png)
+
+**Katalon Studio** is an excellent end-to-end automation solution for web, API, mobile, and desktop testing with DevOps support.
+
+With Katalon Studio, automated testing can be easily integrated into any CI/CD pipeline to release products faster while guaranteeing high quality. Katalon Studio customizes for users from beginners to experts. Robust functions such as Spying, Recording, Dual-editor interface and Custom Keywords make setting up, creating and maintaining tests possible for users.
+
+Built on top of Selenium and Appium, Katalon Studio helps standardize your end-to-end tests standardized. It also complies with the most popular frameworks to work seamlessly with other tools in the automated testing ecosystem.
+
+Katalon is endorsed by Gartner, IT professionals, and a large testing community.
+
+>Note: At the time of this writing, Katalon Studio extension for AzDO was NOT available for Linux.
+
+=> [Visit Katalon Studio Website](https://www.katalon.com/) or read about its [integration with AzDO](https://docs.katalon.com/katalon-studio/docs/azure-devops-extension.html#installation)
+
+## Conclusion
+
+Hope you learned various aspects of E2E testing like its processes, metrics, the difference between Unit, Integration  and E2E testing, and the various recommended E2E test frameworks and tools.
+
+For any commercial release of the software, E2E test verification plays an important role as it tests the entire application in an environment that exactly imitates real-world users like network communication, middleware and backend services interaction, etc.
+
+Finally, the E2E test is often performed manually as the cost of automating such test cases is too high to be afforded by any organization.
```

Comment:
I think the article bets heavily on automation, as it should. More can be said here about manual testing being quite pricey, too, and slow, error prone (if much more clever than a 'dumb script'). There are trade-offs and realities.

Reply:
True. But in reality, the e2e testing often is too complex and expensive to make it fully automated. It is also true, that some parts of the e2e testing can be self-automated, thus the tools listed in the article that come to the rescue.